### PR TITLE
Server Deriving Tokens on behalf of Clients

### PR DESCRIPTION
**Pull request by dadgar**
This PR adds:

* An endpoint for clients to hit to derive tokens
* Raft/FSM handling of VaultAccessors
* VaultAccessor MemDB table

This PR does *NOT* handle token revocation

@armon @diptanu @vishalnayak  for review

### CODE REVIEW

File: nomad/fsm.go [hunk 1]
```diff
@@ -945,6 +977,30 @@ func (s *nomadSnapshot) persistJobSummaries(sink raft.SnapshotSink,
...
+	for {
+		raw := accessors.Next()
+		if raw == nil {
+			break
+		}
+
+		accessor := raw.(*structs.VaultAccessor)
+
+		sink.Write([]byte{byte(VaultAccessorSnapshot)})
+		if err := encoder.Encode(accessor); err != nil {
```
Comment (vishalnayak) [inline nomad/fsm.go]: Can we ensure that `encoder` is non-nil?

File: nomad/fsm.go [hunk 2]
```diff
@@ -454,6 +457,22 @@ func (n *nomadFSM) applyReconcileSummaries(buf []byte, index uint64) interface{}
 	return n.reconcileQueuedAllocations(index)
 }
 
+// applyUpsertVaultAccessor stores the Vault accessors for a given
```
Comment (vishalnayak) [inline nomad/fsm.go]: Comment is accidentally truncated?

File: nomad/fsm.go (hunk 1 above)
Comment (dadgar) [inline nomad/fsm.go]: There is no need as the `Persist` method creates the encoder and returns an error if it fails to write

**Comment by diptanu** [comment]
LGTM overall.

**Pull request merged**